File: solutions/go/anagram/1/anagram.go
package anagram

import (
	"slices"
	"strings"
)

func Detect(subject string, candidates []string) (anagrams []string) {
	sorted := func(s string) []rune {
		runes := []rune(s)
		slices.Sort(runes)
		return runes
	}

	subjectLower := strings.ToLower(subject)
	subjectSorted := sorted(subjectLower)

	isAnagram := func(condidate string) bool {
		condidateLower := strings.ToLower(condidate)
		return condidateLower != subjectLower &&
			slices.Equal(sorted(condidateLower), subjectSorted)
	}

	for _, candidate := range candidates {
		if isAnagram(candidate) {
			anagrams = append(anagrams, candidate)
		}
	}
	return
}

File: solutions/haskell/anagram/1/src/Anagram.hs
{-# LANGUAGE ViewPatterns #-}

module Anagram (anagramsFor) where

import Data.Char (toLower)
import Data.List (sort)

anagramsFor :: String -> [String] -> [String]
anagramsFor w = filter isAnagram
  where
    wl = map toLower w
    ws = sort wl
    isAnagram (map toLower -> cl) = cl /= wl && sort cl == ws
